### PR TITLE
Async transfer batch and block mask to cuda device

### PR DESCRIPTION
**Pull request by apaz-cli**

### CODE REVIEW

File: src/zeroband/data.py [hunk 1]
```diff
@@ -273,13 +275,125 @@ def load_state_dict(self, state_dict):
             dataset.load_state_dict(state_dict[f"dataset_{i}"])
         self._init_random_state()
 
+
+class PrefetchDataLoader(StatefulDataLoader):
```
Comment (samsja) [inline src/zeroband/data.py:279]: can you add docstring explaning how the whole thing is working ?

File: src/zeroband/data.py [hunk 2]
```diff
@@ -273,13 +275,125 @@ def load_state_dict(self, state_dict):
...
+                # Construct and return processed batch.
+                self.ready_batch = {
+                    "input_ids": input_ids,
+                    "labels": labels,
+                    "block_mask": block_mask
+                }
+                return None
+
+            self.thread = threading.Thread(target=_task)
+            self.thread.start()
```
Comment (samsja) [inline src/zeroband/data.py:375]: I am not sure to understand why we need thread in the datalaoder

File: src/zeroband/data.py [hunk 3]
```diff
@@ -273,13 +275,125 @@ def load_state_dict(self, state_dict):
             dataset.load_state_dict(state_dict[f"dataset_{i}"])
         self._init_random_state()
 
+
+class PrefetchDataLoader(StatefulDataLoader):
+    def __init__(self, original_dataloader: StatefulDataLoader, config: Config):
+        self.config = config
```
Comment (samsja) [inline src/zeroband/data.py]: hmm why do we have to pass the top Config ? Ideally would like to avoid this situation to not create convolutated code

File: src/zeroband/data.py (hunk 1 above)
Comment (apaz-cli) [reply src/zeroband/data.py:279]: Yeah, will do.

File: src/zeroband/data.py (hunk 2 above)
Comment (apaz-cli) [reply src/zeroband/data.py:375]: We spin a thread to transfer the next batch to GPU during forward(). This would not be as big a deal, but creating the block mask actually depends on the data in the batch. So it can't start until it's transferred. Also the block mask gets torch compiled, which takes a bunch of cpu time.

Each of these things took about .15 to .2 seconds. So to not wait on those two things for .3 to .4 seconds on every minibatch, we put it on another thread.

File: src/zeroband/data.py (hunk 3 above)
Comment (apaz-cli) [reply src/zeroband/data.py]: We need it so we can honor `config.train.sequence_packing`. Open to other solutions though.

**Review by samsja — APPROVED** [review]

File: src/zeroband/models/llama/model.py [hunk 4]
```diff
@@ -459,7 +459,9 @@ def __init__(self, model_args: ModelArgs):
         self.norm = build_norm(model_args.norm_type, dim=model_args.dim, eps=model_args.norm_eps)
 
         self.output = nn.Linear(model_args.dim, model_args.vocab_size, bias=False)
-        self.init_weights()
+
+        with torch.no_grad():
+            self.init_weights()
```
Comment (samsja) [inline src/zeroband/models/llama/model.py]: I don't see the direct link to this PR. can we do it else where ?

File: src/zeroband/data.py [hunk 5]
```diff
@@ -273,13 +275,132 @@ def load_state_dict(self, state_dict):
...
+        def __del__(self):
+            self._await_prefetch()
+
 def get_dataloader(
     tokenizer,
     world_size: int,
     rank: int,
     batch_size: int,
-    data_config: DataConfig,
+    config: Config,
```
Comment (samsja) [inline src/zeroband/data.py]: lets keep data_config please and avoid top to bottom config passing

File: src/zeroband/data.py [hunk 6]
```diff
@@ -273,13 +275,132 @@ def load_state_dict(self, state_dict):
...
+
+class PrefetchDataLoader(StatefulDataLoader):
+    """
+    This class is a wrapper around a dataloader that prefetches the next batch asynchronously on another thread.
+    This is useful to hide the latency of transferring the batch to GPU.
+    We can't integrate this into the StatefulDataloader's collate_fn() because it runs in another process.
+    We're also using it to hide the latency of torch compiling FlexAttention block masks.
+    """
+
+    def __init__(self, original_dataloader: StatefulDataLoader, config: Config):
```
Comment (samsja) [inline src/zeroband/data.py]: lets stick to data_config here

File: src/zeroband/models/llama/model.py (hunk 4 above)
Comment (apaz-cli) [reply src/zeroband/models/llama/model.py]: Removed

File: src/zeroband/data.py (hunk 6 above)
Comment (apaz-cli) [reply src/zeroband/data.py]: Rewrote some tests that depended on train.sequence_packing rather than data.sequence_packing.

File: src/zeroband/data.py (hunk 5 above)
Comment (apaz-cli) [reply src/zeroband/data.py]: Done, rewrote

**Comment by apaz-cli** [comment]
Should be ready now.

**Comment by apaz-cli** [comment]
Ran 5 steps, both using the same `./scripts/simulate_multi_node_diloco.sh 1 8 src/zeroband/train.py @ configs/10B/H100_devel.toml` on an SXM5 8xH100 machine. The improvement is fairly consistent.

```hs
Before:
01:14:41 [INFO] [Rank 0] step: 5, loss: 14.8759, tokens_per_second: 26436.06, mfu: 26.06

After:
00:55:55 [INFO] [Rank 0] step: 5, loss: 14.8676, tokens_per_second: 29211.94, mfu: 28.80
```
So, not only was the MFU higher, it was consistently higher to the point where it saved ~20 seconds over 5 steps.

The low MFU is because it's doing unoptimized CPU offload even though we have the vRAM not to.
